File: Source/DotNET/Engine.Specs/for_ScreenTemplateResolver/when_resolving_against_the_shared_fixture_corpus.cs
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Text.Json;
using Cratis.Scene.Engine.Screens;
using Cratis.Scene.Model.Layouts;
using Cratis.Scene.Model.Screens;

namespace Cratis.Scene.Engine.for_ScreenTemplateResolver;

public class when_resolving_against_the_shared_fixture_corpus : Specification
{
    record FixtureCase(
        string Name,
        Layout Layout,
        IReadOnlyList<ScreenTemplate> Templates,
        IReadOnlyList<string> ExpectedPlacements,
        IReadOnlyList<string> ExpectedUnplaced,
        IReadOnlyList<string> ExpectedCycles);

    List<FixtureCase> _cases = null!;
    List<(FixtureCase Case, ScreenTemplateResolution Resolution)> _results = null!;

    void Establish()
    {
        using var document = JsonDocument.Parse(File.ReadAllText(Path.Combine(FindRepositoryRoot(), "screen-template-fixtures.json")));
        _cases = [.. document.RootElement.GetProperty("cases").EnumerateArray().Select(ToFixtureCase)];
    }

    void Because() => _results = [.. _cases.Select(fixtureCase => (fixtureCase, ScreenTemplateResolver.Resolve(fixtureCase.Layout, fixtureCase.Templates)))];

    [Fact]
    void should_place_every_template_where_the_corpus_expects() =>
        AssertEach(
            resolution => [.. resolution.Placements.Select(placement => $"{placement.Template}@{placement.Slot}->{placement.Container}#{placement.Depth}")],
            fixtureCase => fixtureCase.ExpectedPlacements);

    [Fact]
    void should_report_the_expected_unplaced_templates() =>
        AssertEach(
            resolution => [.. resolution.Unplaced.Select(unplaced => $"{unplaced.Template}@{unplaced.Slot}?{string.Join('|', unplaced.Candidates)}")],
            fixtureCase => fixtureCase.ExpectedUnplaced);

    [Fact]
    void should_report_the_expected_cycles() =>
        AssertEach(
            resolution => [.. resolution.Cycles.Select(cycle => string.Join('>', cycle))],
            fixtureCase => fixtureCase.ExpectedCycles);

    [Fact]
    void should_consider_a_case_with_nothing_wrong_valid()
    {
        foreach (var (fixtureCase, resolution) in _results)
        {
            var expectedValid = fixtureCase.ExpectedUnplaced.Count == 0 && fixtureCase.ExpectedCycles.Count == 0;
            (fixtureCase.Name, resolution.IsValid).ShouldEqual((fixtureCase.Name, expectedValid));
        }
    }

    void AssertEach(Func<ScreenTemplateResolution, IReadOnlyList<string>> actual, Func<FixtureCase, IReadOnlyList<string>> expected)
    {
        foreach (var (fixtureCase, resolution) in _results)
        {
            (fixtureCase.Name, string.Join(',', actual(resolution))).ShouldEqual((fixtureCase.Name, string.Join(',', expected(fixtureCase))));
        }
    }

    static FixtureCase ToFixtureCase(JsonElement element) =>
        new(
            element.GetProperty("name").GetString()!,
            ToLayout(element.GetProperty("layout")),
            [.. element.GetProperty("templates").EnumerateArray().Select(ToTemplate)],
            [.. element.GetProperty("expectedPlacements").EnumerateArray().Select(placement =>
                $"{placement.GetProperty("template").GetString()}@{placement.GetProperty("slot").GetString()}->{placement.GetProperty("container").GetString()}#{placement.GetProperty("depth").GetInt32()}")],
            [.. element.GetProperty("expectedUnplaced").EnumerateArray().Select(unplaced =>
                $"{unplaced.GetProperty("template").GetString()}@{unplaced.GetProperty("slot").GetString()}?{string.Join('|', unplaced.GetProperty("candidates").EnumerateArray().Select(candidate => candidate.GetString()))}")],
            [.. element.GetProperty("expectedCycles").EnumerateArray().Select(cycle => string.Join('>', cycle.EnumerateArray().Select(name => name.GetString()!)))]);

    static Layout ToLayout(JsonElement element) =>
        new(element.GetProperty("name").GetString()!, [.. Slots(element)]);

    static ScreenTemplate ToTemplate(JsonElement element) =>
        new(
            element.GetProperty("name").GetString()!,
            element.GetProperty("fitsSlot").ValueKind == JsonValueKind.Null ? null : element.GetProperty("fitsSlot").GetString(),
            [.. Slots(element)]);

    static IEnumerable<Slot> Slots(JsonElement element) =>
        element.GetProperty("slots").EnumerateArray().Select(slot => new Slot(slot.GetString()!));

    static string FindRepositoryRoot()
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        while (directory is not null && !File.Exists(Path.Combine(directory.FullName, "Scene.slnx")))
        {
            directory = directory.Parent;
        }

        return directory?.FullName ?? throw new DirectoryNotFoundException("Could not locate the repository root (Scene.slnx) above " + AppContext.BaseDirectory);
    }
}

File: Source/JavaScript/engine/for_resolveScreenTemplates/when_resolving_against_the_shared_fixture_corpus.ts
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { Layout, ScreenTemplate } from '@cratis/scene.model';
import { isScreenTemplateResolutionValid, resolveScreenTemplates } from '../index';

interface FixtureCase {
    name: string;
    layout: { name: string; slots: string[] };
    templates: { name: string; fitsSlot: string | null; slots: string[] }[];
    expectedPlacements: { template: string; slot: string; container: string; depth: number }[];
    expectedUnplaced: { template: string; slot: string; candidates: string[] }[];
    expectedCycles: string[][];
}

const fixturePath = join(import.meta.dirname, '..', '..', '..', '..', 'screen-template-fixtures.json');
const corpus = JSON.parse(readFileSync(fixturePath, 'utf-8')) as { cases: FixtureCase[] };

describe('when resolving against the shared fixture corpus', () => {
    for (const fixtureCase of corpus.cases) {
        const layout: Layout = { name: fixtureCase.layout.name, slots: fixtureCase.layout.slots.map((name) => ({ name })) };
        const templates: ScreenTemplate[] = fixtureCase.templates.map((template) => ({
            name: template.name,
            fitsSlot: template.fitsSlot ?? undefined,
            slots: template.slots.map((name) => ({ name })),
        }));

        const resolution = resolveScreenTemplates(layout, templates);

        it(`should place every template where the corpus expects for "${fixtureCase.name}"`, () => {
            resolution.placements.should.deep.equal(fixtureCase.expectedPlacements);
        });

        it(`should report the expected unplaced templates for "${fixtureCase.name}"`, () => {
            resolution.unplaced.should.deep.equal(fixtureCase.expectedUnplaced);
        });

        it(`should report the expected cycles for "${fixtureCase.name}"`, () => {
            resolution.cycles.should.deep.equal(fixtureCase.expectedCycles);
        });

        it(`should consider "${fixtureCase.name}" valid only when nothing is wrong`, () => {
            const expectedValid = fixtureCase.expectedUnplaced.length === 0 && fixtureCase.expectedCycles.length === 0;
            isScreenTemplateResolutionValid(resolution).should.equal(expectedValid);
        });
    }
});
